[Dlight] Benchmarking Tools for Dynamic Shape PrimFuncs & Relax Function

zxybazh opened the pull request:
This PR introduces benchmarking tools for dynamic shape PrimFuncs and Relax. It facilitates the following functionalities:
1. Automatically benchmark PrimFunc performance with user-speicified dynamic shape sampling function and input information. E.g., `n=random.randint(50, 100)` will benchmark performance of PrimFunc with Dynamic dimension `n` between 50-100.
2. Extract self-contained PrimFunc benchmarking files from Relax Module. E.g., it can produce multiple python files for each function in a Relax Module and automatically extract dynamic shape input information from bindings.
3. Conduct Relax Function level benchmarking using the same dynamic shape sample value across the Relax Function. E.g., the same value `n` is used consistently in a Relax Function for any PrimFunction call so we can figure out what is the performance bottleneck when we pin down the value of all dynamic dimensions like `n` or `m` (Yes, there could be multiple dynamic dimensions).
4. It can automatically generate valid input even when the input is not the dynamic shape but the value of the dynamic dimension, which is `n`. This is specific to `rotatry_embedding` for now.

Co-authored by @cyx-6
Thanks to @junrushao for advices.

Example Usage
```python
def benchmark_prim_func_full_rpc():
    with LocalRPC() as rpc:
        rpc_config = ms.runner.RPCConfig(
            tracker_host=rpc.tracker_host,
            tracker_port=rpc.tracker_port,
            tracker_key=rpc.tracker_key,
            session_priority=1,
            session_timeout_sec=100,
        )
        benchmark_prim_func(
            cuda_workload,
            args=[
                ((1, "m", 4096), "float32"),
                ((4096, 4096), "float32"),
                ((1, "m", 4096), "float32"),
            ],
            dym_var_dict={"m": "int32"},
            target="nvidia/geforce-rtx-3070",
            dev=tvm.cuda(),
            rpc_config=rpc_config,
            evaluator_config=ms.runner.EvaluatorConfig(
                number=10,
                repeat=10,
                min_repeat_ms=0,
                enable_cpu_cache_flush=False,
            ),
        )
```

Expected Results for the tested PrimFunc:
```
  InputInfo   Time(us)    Std(us)  Weight  WxTime(ms)
0   m = 126  752.48000  19.417429       1    0.752480
1    m = 56  430.68955   0.244274       1    0.430690
2    m = 13  340.11350   0.241286       1    0.340114
3    m = 89  692.58875   0.343988       1    0.692589
4    m = 98  819.43990   0.316655       1    0.819440
```

sunggg commented on tests/python/dlight/test_benchmark.py:
```diff
@@ -0,0 +1,191 @@
+        dev=tvm.cuda(),
+    )
+
+
```
Since this PR brings `benchmark_relax_func`, shall we add its testcase?

sunggg commented on tests/python/dlight/test_benchmark.py:
```diff
@@ -0,0 +1,191 @@
+        target="nvidia/geforce-rtx-3070",
+        dev=tvm.cuda(),
+    )
+
```
Also, some testcases for extraction would be helpful.

sunggg reviewed (commented):
Great work, @zxybazh @cyx-6! A few minor comments.

zxybazh commented:
Thanks @sunggg for reviewing! The corresponding tests has been added, would you please take another look?

Some tests require cuda so I've run locally to test, and one test case using cpu was also added. For extraction, it's a bit hard to verify the results so I added a test to run the flow and manually checked offline. The result was pretty good to use. Here's a sample of the automated extracted python script:

```python
import pickle

import tvm
from tvm import relax
from tvm.script import tir as T

from tvm.dlight.benchmark import benchmark_prim_func

MODEL_NAME = "TEST"
RELAX_FUNC_NAME = "test"
PRIM_FUNC_NAME = "matmul1"
FUNC_HASH = -7013748748655302229
WEIGHT = 2
SAMPLE_NUMBER = 10

INPUT_ARGS = pickle.loads(b'\x80\x05\x95<\x00\x00\x00\x00\x00\x00\x00]\x94((K\x01K K\x01\x8c\x01n\x94t\x94\x8c\x07float16\x94\x86\x94(K\x01K h\x01K\x80t\x94h\x03\x86\x94(K\x01K K\x01K\x80t\x94h\x03\x86\x94e.')
DYM_VAR_SAMPLE_FUNC = pickle.loads(b'\x80\x05\x95>\x00\x00\x00\x00\x00\x00\x00\x8c\x1atvm.dlight.benchmark.utils\x94\x8c\x1bdefault_dym_var_sample_func\x94\x93\x94.')
DYM_VAR_DICT = pickle.loads(b'\x80\x05\x95\x10\x00\x00\x00\x00\x00\x00\x00}\x94\x8c\x01n\x94\x8c\x05int32\x94s.')

# from tvm.script import tir as T

@T.prim_func
def main(var_A: T.handle, var_B: T.handle, matmul: T.Buffer((T.int64(1), T.int64(32), T.int64(1), T.int64(128)), "float16")):
    T.func_attr({"op_pattern": 4, "tir.noalias": T.bool(True)})
    n = T.int64()
    A = T.match_buffer(var_A, (T.int64(1), T.int64(32), T.int64(1), n), "float16")
    B = T.match_buffer(var_B, (T.int64(1), T.int64(32), n, T.int64(128)), "float16")
    # with T.block("root"):
    for i0, i1, i2, i3, k in T.grid(T.int64(1), T.int64(32), T.int64(1), T.int64(128), n):
        with T.block("matmul"):
            v_i0, v_i1, v_i2, v_i3, v_k = T.axis.remap("SSSSR", [i0, i1, i2, i3, k])
            T.reads(A[v_i0, v_i1, v_i2, v_k], B[v_i0, v_i1, v_k, v_i3])
            T.writes(matmul[v_i0, v_i1, v_i2, v_i3])
            with T.init():
                matmul[v_i0, v_i1, v_i2, v_i3] = T.float16(0)
            matmul[v_i0, v_i1, v_i2, v_i3] = matmul[v_i0, v_i1, v_i2, v_i3] + A[v_i0, v_i1, v_i2, v_k] * B[v_i0, v_i1, v_k, v_i3]

if __name__ == "__main__":
    target = tvm.target.Target("llvm -num-cores=4")
    dev = tvm.cpu()
    print("Input args:", INPUT_ARGS)
    benchmark_prim_func(
        main,
        args=INPUT_ARGS,
        dym_var_dict=DYM_VAR_DICT,
        dym_var_sample_func=DYM_VAR_SAMPLE_FUNC,
        sample_number=SAMPLE_NUMBER,
        target=target,
        dev = dev,
        weight = WEIGHT,
        relax_func_name = RELAX_FUNC_NAME,
        prim_func_name = PRIM_FUNC_NAME,
    )

```

zxybazh commented:
Hi @sunggg automatic input argument extraction from the PrimFunc has been added following @Hzfengsy 's colab notebook, would you please take another look?

sunggg reviewed (approved):
LGTM! Thank you :)
